### DISCUSSION TITLE
How to get member uvalue of Derived class, from Base*

### DISCUSSION THREAD

**Discussion by MineBill** [BlackMATov/meta.hpp#88] [Q&A] [answered]
Hello,

I have a vector of Base* and using resolve_type, i can successfully get members of Derived classes but when checking with `member.is_gettable_with` using that Base*, or a Base& it's always false. It'll only work when providing the type of the Derived class(_using `static_cast<DerivedA*>(base_ptr)`_) but at that point, it's not so dynamic any more. Is this not supported yet or am i missing something. I went through the docs multiple times but i couldn't find anything.

In any case, thank you so much for this library!

**Comment by MineBill** [comment]
After some investigation, it looks like `resolve_type` and `ucast` are the only places that take into account that the type might have poly info. Everything else, probably `get` and `is_gettable_with` too, will only check with the `Base` and ignore poly info.

**Comment by BlackMATov** [comment]
Hello and thanks for your interest in this library!

So, yes, you made a good research. We don't have enough dynamic functions to manage your case.

We can't using instance's poly info in functions like `is_gettable_with` or `is_invocable_with`, because it will cause a lot of problems with multiple inheritance, virtual functions, etc. This functions can only rely on the current type of the instance and use upcasts as regular C++ does.

To resolve your problem, I think we should have an additional `ucast` with dynamic type destination. It can look like this:

```
template < typename From >
uvalue ucast(From&& from, const class_type& to);
```

Where `From` is a pointer or `uvalue` with a pointer. Also, we have to return `uvalue` because we don't know static `To` type.

I believe with this, you can cast your base instance to the most derived class and call whatever you want.

I'm going to think about it. And I will be glad to any suggestions!

**Comment by MineBill** [comment]
Thank you for taking time to comment!

So, with this additional `ucast` i should be able to do something like this:
```cpp
...
class_type klass = resolve_type(base_ptr);
uvalue var = klass.get_member("var").get(ucast(base_ptr, klass));
...
```

Honestly, i would be more than happy if i could just get a `void*` to the field/function.

To give you more context, what I'm trying to achieve is to get pointers to members of component classes that derive from a base `Component` class and give them to the appropriate `ImGui::Input*` function in [imgui](https://github.com/ocornut/imgui). The only "missing" part is getting that pointer. I believe the `member_type` gives me enough information to figure out the actual type, which i can then cast the pointer to before calling into imgui.

**Comment by BlackMATov** [comment, marked as answer]
Added new `meta_poly_ptr` method to `META_HPP_ENABLE_POLY_INFO` macro.

It returns a pointer of the most derived type, which can be used as an instance for member access. See a little example with comments here: https://github.com/BlackMATov/meta.hpp/blob/main/develop/untests/meta_issues/github_discussion_88.cpp

I believe it should help! :-)

**Comment by MineBill** [comment]
Wow! Thank you so much for implementing this so quickly! It works as expected now! Thanks!
